File: npmjs/pray-cli/src/cli/invocation.ts
import { PrayError } from "../errors.js";
import {
  activeInvocationContext,
  lockfilePathForContext,
  projectInvocationContextFromCurrentDirectory,
  projectInvocationContextFromOptions,
  setActiveInvocationContext,
  type ProjectInvocationContext,
  type ProjectInvocationOptions,
} from "../project-context/index.js";
import { defaultResolveOptions, type ResolveOptions } from "../resolve/context.js";
import {
  resolveProject,
  resolveProjectWithGitRefreshFallback,
} from "../resolve/project.js";
import type { ResolvedProject } from "../resolve/types.js";

export function initializeInvocation(argumentsList: string[]): string[] {
  const { options, remaining } = parseGlobalOptions(argumentsList);
  const context = projectInvocationContextFromOptions(options);
  setActiveInvocationContext(context);
  return remaining;
}

export function manifestPath(): string {
  return invocationContextValue().manifestPath;
}

export function projectRoot(): string {
  return invocationContextValue().projectRoot;
}

export function lockfilePath(): string {
  return lockfilePathForContext(invocationContextValue());
}

export function selectedEnvironment(): string | undefined {
  return invocationContextValue().environment;
}

export async function resolveCurrentProject(
  options: ResolveOptions = defaultResolveOptions(),
): Promise<ResolvedProject> {
  const context = invocationContextValue();
  const resolveOptions: ResolveOptions = {
    ...options,
    environment: options.environment ?? context.environment,
  };
  return resolveProject(context.manifestPath, resolveOptions);
}

export async function resolveCurrentProjectWithGitRefreshFallback(
  options: ResolveOptions = defaultResolveOptions(),
  allowGitRefreshFallback = false,
): Promise<ResolvedProject> {
  const context = invocationContextValue();
  const resolveOptions: ResolveOptions = {
    ...options,
    environment: options.environment ?? context.environment,
  };
  return resolveProjectWithGitRefreshFallback(
    context.manifestPath,
    resolveOptions,
    allowGitRefreshFallback,
  );
}

function invocationContextValue(): ProjectInvocationContext {
  return (
    activeInvocationContext() ?? projectInvocationContextFromCurrentDirectory()
  );
}

function parseGlobalOptions(argumentsList: string[]): {
  options: ProjectInvocationOptions;
  remaining: string[];
} {
  const options: ProjectInvocationOptions = {};
  const remaining: string[] = [];
  let index = 0;
  while (index < argumentsList.length) {
    const argument = argumentsList[index]!;
    switch (argument) {
      case "--path":
        options.projectRoot = requireOptionValue("--path", argumentsList[++index]);
        index++;
        continue;
      case "--file-path":
        options.manifestPath = requireOptionValue(
          "--file-path",
          argumentsList[++index],
        );
        index++;
        continue;
      case "--env":
      case "--environment":
        options.environment = requireEnvironmentValue(
          argument,
          argumentsList[++index],
        );
        index++;
        continue;
      default:
        if (isTopLevelCommand(argument)) {
          remaining.push(...argumentsList.slice(index));
          return { options, remaining };
        }
        if (argument.startsWith("-")) {
          remaining.push(...argumentsList.slice(index));
          return { options, remaining };
        }
        remaining.push(...argumentsList.slice(index));
        return { options, remaining };
    }
  }
  return { options, remaining };
}

function requireOptionValue(flag: string, value: string | undefined): string {
  if (value === undefined) {
    throw PrayError.usage(`${flag} requires a value`);
  }
  return value;
}

function requireEnvironmentValue(
  flag: string,
  value: string | undefined,
): string {
  if (value === undefined) {
    throw PrayError.usage(`${flag} requires a value`);
  }
  return value;
}

function isTopLevelCommand(token: string): boolean {
  return [
    "manifest",
    "init",
    "prayer",
    "repo",
    "install",
    "add",
    "remove",
    "update",
    "unlock",
    "render",
    "plan",
    "apply",
    "verify",
    "drift",
    "format",
    "package",
    "publish",
    "login",
    "serve",
    "confess",
    "list",
    "outdated",
    "explain",
    "vendor",
    "clean",
    "tree",
    "sync",
    "trust",
    "version",
    "-V",
    "--version",
    "-h",
    "--help",
  ].includes(token);
}

File: crates/pray-cli/src/invocation.rs
use pray_core::project_context::{ProjectInvocationContext, ProjectInvocationOptions};
use pray_core::resolve::{resolve_project_in_context, ResolvedProject};
use pray_core::resolve_context::ResolveOptions;
use pray_core::{PrayError, PrayResult};
use std::cell::RefCell;
use std::path::PathBuf;

thread_local! {
    static INVOCATION: RefCell<Option<ProjectInvocationContext>> = const { RefCell::new(None) };
}

pub fn initialize(arguments: Vec<String>) -> PrayResult<Vec<String>> {
    let (options, remaining) = parse_global_options(arguments)?;
    let context = ProjectInvocationContext::from_options(options)?;
    INVOCATION.with(|cell| {
        *cell.borrow_mut() = Some(context);
    });
    Ok(remaining)
}

pub fn manifest_path() -> PathBuf {
    invocation_context().manifest_path
}

pub fn lockfile_path() -> PathBuf {
    invocation_context().lockfile_path()
}

pub fn project_root() -> PathBuf {
    invocation_context().project_root
}

pub fn resolve_current_project(options: &ResolveOptions) -> PrayResult<ResolvedProject> {
    let context = invocation_context();
    let mut options = options.clone();
    if options.environment.is_none() {
        options.environment = context.environment.clone();
    }
    resolve_project_in_context(&context.manifest_path, &context.project_root, &options)
}

fn invocation_context() -> ProjectInvocationContext {
    INVOCATION.with(|cell| {
        cell.borrow().clone().unwrap_or_else(|| {
            ProjectInvocationContext::from_current_directory().expect("project invocation context")
        })
    })
}

fn parse_global_options(
    arguments: Vec<String>,
) -> PrayResult<(ProjectInvocationOptions, Vec<String>)> {
    let mut options = ProjectInvocationOptions::default();
    let mut remaining = Vec::new();
    let mut iter = arguments.into_iter();
    while let Some(argument) = iter.next() {
        match argument.as_str() {
            "--path" => {
                options.project_root = Some(require_option_value("--path", iter.next())?);
            }
            "--file-path" => {
                options.manifest_path = Some(require_option_value("--file-path", iter.next())?);
            }
            "--env" | "--environment" => {
                options.environment = Some(require_environment_value(&argument, iter.next())?);
            }
            value if is_top_level_command(value) => {
                remaining.push(argument);
                remaining.extend(iter);
                break;
            }
            other if other.starts_with('-') => {
                remaining.push(argument);
                remaining.extend(iter);
                break;
            }
            _ => {
                remaining.push(argument);
                remaining.extend(iter);
                break;
            }
        }
    }
    Ok((options, remaining))
}

fn require_option_value(flag: &str, value: Option<String>) -> PrayResult<PathBuf> {
    value
        .map(PathBuf::from)
        .ok_or_else(|| PrayError::Usage(format!("{flag} requires a value")))
}

fn require_environment_value(flag: &str, value: Option<String>) -> PrayResult<String> {
    value.ok_or_else(|| PrayError::Usage(format!("{flag} requires a value")))
}

fn is_top_level_command(token: &str) -> bool {
    matches!(
        token,
        "manifest"
            | "init"
            | "prayer"
            | "repo"
            | "install"
            | "add"
            | "remove"
            | "update"
            | "unlock"
            | "render"
            | "plan"
            | "apply"
            | "verify"
            | "drift"
            | "format"
            | "package"
            | "publish"
            | "login"
            | "serve"
            | "confess"
            | "list"
            | "outdated"
            | "explain"
            | "vendor"
            | "clean"
            | "tree"
            | "sync"
            | "trust"
            | "version"
            | "-V"
            | "--version"
            | "-h"
            | "--help"
    )
}
